From 000064171db044f98d1c9e025e16007d8269e26a Mon Sep 17 00:00:00 2001
From: Orko Garai <orko@igalia.com>
Date: Fri, 20 Sep 2024 16:05:11 +0000
Subject: [PATCH] [ozone/wayland] Check format supported by gbm for creating native pixmap

When creating native pixmap without handle,
OzonePlatform::IsNativePixmapConfigSupported should return false in
wayland if gbm does not support a format. This is because gbm is the
only method used when creating native pixmap without handle.

Also check if format is supported by gbm in
WaylandSurfaceFactory::CreateNativePixmap().

Fixed: 331796411, 365399706
Change-Id: I522601f5872e5581d5e7d8acf83419c8b2881ab7
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5873796
Reviewed-by: Kramer Ge <fangzhoug@chromium.org>
Commit-Queue: Orko Garai <orko@igalia.com>
Cr-Commit-Position: refs/heads/main@{#1358193}
---

diff --git a/ui/ozone/platform/wayland/gpu/wayland_surface_factory.cc b/ui/ozone/platform/wayland/gpu/wayland_surface_factory.cc
index bab048c4..0e0ac92e 100644
--- a/ui/ozone/platform/wayland/gpu/wayland_surface_factory.cc
+++ b/ui/ozone/platform/wayland/gpu/wayland_surface_factory.cc
@@ -250,17 +250,20 @@
     return nullptr;
   }
 #if defined(WAYLAND_GBM)
-  scoped_refptr<GbmPixmapWayland> pixmap =
-      base::MakeRefCounted<GbmPixmapWayland>(buffer_manager_);
+  auto* gbm_device = buffer_manager_->GetGbmDevice();
+  if (gbm_device->CanCreateBufferForFormat(
+          GetFourCCFormatFromBufferFormat(format))) {
+    scoped_refptr<GbmPixmapWayland> pixmap =
+        base::MakeRefCounted<GbmPixmapWayland>(buffer_manager_);
 
-  if (!pixmap->InitializeBuffer(widget, size, format, usage,
-                                framebuffer_size)) {
-    return nullptr;
+    if (!pixmap->InitializeBuffer(widget, size, format, usage,
+                                  framebuffer_size)) {
+      return nullptr;
+    }
+    return pixmap;
   }
-  return pixmap;
-#else
-  return nullptr;
 #endif
+  return nullptr;
 }
 
 void WaylandSurfaceFactory::CreateNativePixmapAsync(
diff --git a/ui/ozone/platform/wayland/ozone_platform_wayland.cc b/ui/ozone/platform/wayland/ozone_platform_wayland.cc
index bcb8d50e..aebe9a8 100644
--- a/ui/ozone/platform/wayland/ozone_platform_wayland.cc
+++ b/ui/ozone/platform/wayland/ozone_platform_wayland.cc
@@ -77,6 +77,8 @@
 #endif
 
 #if defined(WAYLAND_GBM)
+#include "ui/gfx/linux/drm_util_linux.h"
+#include "ui/gfx/linux/gbm_device.h"
 #include "ui/ozone/platform/wayland/gpu/drm_render_node_path_finder.h"
 #endif
 
@@ -204,6 +206,15 @@
       if (!buffer_manager_->SupportsFormat(format)) {
         return false;
       }
+      // Return false here if creating buffers for certain formats is not
+      // possible (e.g. YUV formats are not supported by linux system libgbm
+      // gbm_bo_create) even though |buffer_manager_| may indicate it can be
+      // imported as wl_buffer.
+      auto* gbm_device = buffer_manager_->GetGbmDevice();
+      if (!gbm_device->CanCreateBufferForFormat(
+              GetFourCCFormatFromBufferFormat(format))) {
+        return false;
+      }
     } else {
       if (supported_buffer_formats_.find(format) ==
           supported_buffer_formats_.end()) {
